### PR TITLE
(re)Implement logging

### LINKED ISSUE #35: Implement logging
Dependent on #26 

Populate an "input" element with a descriptive "name" tag on student submission with logging information. This functionality should go into `addLogEntry`.

Also include timestamp in log entries

### PR DESCRIPTION
**Pull request by nelson-lojo**
Construct a log from actions performed by the student and store it in a hidden `"log"` input field (i.e. `class="log" name="{{answers-name}}log"`). Content is a JSON string.

This means that log information is accumulated until submission

Log events:
- opening of the problem
- block rearrangements (into and out of output tray, within input/output trays)
- blank edits (filling in/clearing a blank)

### CODE REVIEW

File: pl-faded-parsons.js [hunk 1]
```diff
@@ -783,7 +819,36 @@ class ParsonsWidget {
...
-  addLogEntry() {}
+  /** Add a tagged, timestamped log entry to `this.config.logStorage` */
+  addLogEntry(tag, data) {
+    const timestamp = new Date();
+
+    const entry = {
+      timestamp: timestamp,
+      tag: tag,
+      data: data
+    }
```
Comment (SybelBlue) [inline pl-faded-parsons.js:830]: small thing `{ timestamp, tag, data };` is more js-esque

File: pl-faded-parsons.js [hunk 2]
```diff
@@ -577,10 +607,16 @@ class ParsonsWidget {
   }
   codelineSummary(line, idx) {
     // this schema is used in pl-faded-parsons.py `read_lines`!
```
Comment (SybelBlue) [inline pl-faded-parsons.js:609]: another small thing, could we keep this comment with the return statement so it warns against anyone changing it?

File: pl-faded-parsons.js [hunk 3]
```diff
@@ -523,6 +523,34 @@ class ParsonsWidget {
...
+        $(tray)
+          .find("li.codeline")
+          .each((codelineNumber, codeline) => {
+            if ($(codeline).attr("logging-id")) return;
+            const codelineId = `${trayNumber}.${codelineNumber}`;
+            $(codeline).attr("logging-id", codelineId);
+            findBlanksIn(codeline).each((blankNumber, blank) => {
+              $(blank).attr("logging-id", `${codelineId}.${blankNumber}`);
+
+              $(blank).on({ // logging hooks for blank edits
```
Comment (SybelBlue) [inline pl-faded-parsons.js:541]: Recommend pulling this out of the big nested callback, use `findBlanksIn(widget.config.main).on` instead, give it it's own comment description

File: pl-faded-parsons.js [hunk 4]
```diff
@@ -523,6 +523,34 @@ class ParsonsWidget {
...
+    
+    // add uids to each line and blank for logging ////////////////////////////
+    //  + add logging hooks for blank edits       ////////////////////////////
+    $(widget.config.main)
+      .find(".codeline-tray")
+      .each((trayNumber, tray) =>
+        $(tray)
+          .find("li.codeline")
+          .each((codelineNumber, codeline) => {
+            if ($(codeline).attr("logging-id")) return;
```
Comment (SybelBlue) [inline pl-faded-parsons.js:535]: I know I wrote this line but I don't know if it makes sense in your version of logging. I'm not sure if the system remembers ids between submissions, so I don't know if this return condition is ever met...

**Review by SybelBlue — CHANGES_REQUESTED** [review]
Only a couple small pointers! Otherwise it looks great.

File: pl-faded-parsons.js (hunk 4 above)
Comment (nelson-lojo) [reply pl-faded-parsons.js:534]: Made it so that logs are preserved between submissions, tested it on a dummy course with an FPP

**Comment by nelson-lojo** [comment]
As it stands, logs accumulate between submissions. This is not an issue (just take the last set of logs per variant the student sees), but something to keep in mind

File: pl-faded-parsons.py [hunk 5]
```diff
@@ -222,6 +222,7 @@ def get_code_lines():
     html_params.update({
         "given" : given,
         "uuid": pl.get_uuid(),
+        "previous_log" : data['submitted_answers'].get('log', "[]")
```
Comment (SybelBlue) [inline pl-faded-parsons.py:225]: This adds to the existing bug referenced in #39 - I believe the real dictionary key would be `answers_name + 'log'` when answers_name is set (it should be required but we don't do that yet).

**Review by SybelBlue — APPROVED** [review]
Worsening an existing bug is bad form... but I'm okay with it for now. I'll take a swing #36 when I get a chance and after we get to a stable release for the study.

**Pull request merged**